server/store.py

```diff
@@ -0,0 +1,337 @@
+        # and update the existing index
+        self.update_status(State.UPDATING, "Updating index")
+        try:
+            if source == Source.DAILY:
+                await self.index_daily_recordings()
+            elif source == Source.UPLOADS:
+                await self.generate_upload_transcripts()
+            self.index.storage_context.persist(get_index_dir_path())
+            self.update_status(State.READY, "Index ready to query")
+        except Exception as e:
+            print(
+                f"failed to update index from source {source} - {e}",
+                file=sys.stderr)
+            traceback.print_exc()
+            self.update_status(State.ERROR, "Failed to update existing index")
+
+    async def generate_index(self, source: Source):
+        """Generates a new index from given source"""
+        if source == Source.DAILY:
+            await self.index_daily_recordings()
+        elif source == Source.UPLOADS:
+            await self.generate_upload_transcripts()
+        self.create_index()
+
+    async def generate_upload_transcripts(self):
+        """Generates transcripts from uploaded files"""
+        tasks = []
+
+        # Process five files at a time
+        sem = asyncio.BoundedSemaphore(5)
+        uploaded_file_paths = get_uploaded_file_paths()
+        for path in uploaded_file_paths:
+            if not path.endswith(".mp4") and not path.endswith(".mov"):
+                continue
+            tasks.append(asyncio.create_task(
+                self.index_uploaded_file(sem, path)))
+        # Wait for all processing tasks to finish
+        await asyncio.gather(*tasks)
+
+    def create_index(self):
+        """Creates a new index
+         See: https://gpt-index.readthedocs.io/en/latest/examples/vector_stores/ChromaIndexDemo.html
+        """
+
+        # Get all documents from the present transcripts
+        documents = SimpleDirectoryReader(
+            get_transcripts_dir_path()
+        ).load_data()
+
+        vector_store = self.get_vector_store()
+        storage_context = StorageContext.from_defaults(
+            vector_store=vector_store)
+        index = VectorStoreIndex.from_documents(
+            documents, storage_context=storage_context
+        )
+        index.storage_context.persist(persist_dir=get_index_dir_path())
+        self.index = index
+
+    async def index_daily_recordings(self):
+        """Indexes Daily recordings as per provided room name and max recordings configuration"""
+        recordings = fetch_recordings(self.daily_room_name, self.max_videos)
+
+        # Process up to 5 recordings at a time
+        sem = asyncio.BoundedSemaphore(5)
+        tasks = []
+        for recording in recordings:
+            tasks.append(asyncio.create_task(
+                self.index_daily_recording(sem, recording)))
+        # Wait for tasks to complete
+        await asyncio.gather(*tasks)
+
+    async def index_daily_recording(self, sem: asyncio.locks.BoundedSemaphore,
```

Comment:
I think this is wonky... I am using a bounded semaphore here to limit us to processing 5 recordings at a time, but I have a feeling it may be better to create a single reusable executor instance on this class with a `max_workers` parameter instead. WDYT?

Reply:
this looks neat! but yes probably an executor would be simpler?

```
with ThreadPoolExecutor(max_workers=5) as executor:
        tasks = []
        for recording in recordings:
            tasks.append(asyncio.create_task(self.index_daily_recording(executor, recording)))
        # Wait for tasks to complete
        await asyncio.gather(*tasks)
```

and then:

```
async def index_daily_recording(self, executor: asyncio.concurrent.futures.Executor, recording: Recording):
    loop = asyncio.get_event_loop()
    await loop.run_in_executor(executor, self.transcribe_and_index_recording, recording)
```

I think you need the tasks otherwise you will just be creating one thread (and waiting for it with `await loop.run_in_executor(...)`) at a time instead of 5 and you won't know when they are all finished.

Reply:
I think what can make the task redundant is that I don't need to await each executor run. I could invoke `async.run_in_executor()` as many times as I need and then `async.gather()` them all at the end. I've just pushed the approach I'm proposing there, which should also allow cancellation of futures if server shutdown is invoked as well as enable removal of one extra "nested" level of methods in the store. Does that look reasonable?

Reply:
Ah, yes! Perfect!

Reply:
I didn't realize `run_in_executor` returned a future. :raised_hands: !